From c84be68fe02f88528755dc8f38422c8fbc547217 Mon Sep 17 00:00:00 2001
From: Shivani Soni <quic_soni@quicinc.com>
Date: Sun, 5 Mar 2023 13:52:39 +0530
Subject: [PATCH 320/500] net: bonding: Keep the bond mac address same during
 slave release

Bond netdevice's mac address should not be changed for MLO case

Change-Id: If113d7131dab3e4243a07e8aab133269ebc2465e
Signed-off-by: Shivani Soni <quic_soni@quicinc.com>
---
 drivers/net/bonding/bond_main.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/bonding/bond_main.c b/drivers/net/bonding/bond_main.c
index 5887e9756ae4..1888a068887a 100644
--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -2525,13 +2525,23 @@ static int __bond_release_one(struct net_device *bond_dev,
 	}
 
 	bond_set_carrier(bond);
-	if (!bond_has_slaves(bond))
-		eth_hw_addr_random(bond_dev);
+
+	/*
+	 * Avoid changing the mac address of bond netdevice for MLO case,
+	 * This will only be supported from wifi driver.
+	 */
+	if (BOND_MODE(bond) != BOND_MODE_MLO) {
+		if (!bond_has_slaves(bond))
+			eth_hw_addr_random(bond_dev);
+	}
 
 	unblock_netpoll_tx();
 	synchronize_rcu();
 	bond->slave_cnt--;
 
+	/*
+	 * TODO: Avoid MAC address change notification for BOND_MODE_MLO
+	 */
 	if (!bond_has_slaves(bond)) {
 		call_netdevice_notifiers(NETDEV_CHANGEADDR, bond->dev);
 		call_netdevice_notifiers(NETDEV_RELEASE, bond->dev);
-- 
2.34.1

